### ISSUE TITLE
Help : don't understand the behavior of 2d cast_motion

### ISSUE THREAD

**Issue by finalherizo** [godotengine/godot#17937]
**Godot version:**
3.0.2 win64

**OS/device including version:**
Windows 10 1803 x64 / OpenGL ES 3.0 Renderer: AMD Radeon (TM) R9 380 Series

**Issue description:**
As from the doc, cast_motion should be used to test if a shape can move successfully to a point. I have made many tests but unsuccessful so far. I'm unable to understand the result. Perhaps caused by wrong parameters. This person seems to have the same problem as mine [https://www.reddit.com/r/godot/comments/88cqph/how_does_cast_motion_work/](https://www.reddit.com/r/godot/comments/88cqph/how_does_cast_motion_work/)

**Steps to reproduce:**
Created a root node and 2 bodies. 1 Kinematicbody2D and 1 StaticBody2D respectively with 1 rectangle CollisionShape2D each. 
Added the script below to the CollisionShape2D of the KinematicBody2D.

```gdscript 
extends CollisionShape2D

# class member variables go here, for example:
# var a = 2
# var b = "textvar"
var origin = Vector2()
var dest = Vector2()
var collision_safe = Vector2()
var angle = 0

func _ready():
	# Called every time the node is added to the scene.
	# Initialization here
	pass

func _process(delta):
#	# Called every frame. Delta is time since last frame.
#	# Update game logic here.
	origin = self.position
	dest = origin + (Vector2(1, 0) * 400).rotated(deg2rad(angle))
	
	var space_state = get_world_2d().direct_space_state
	
	var global_origin = global_transform.xform(origin)
	var global_dest = global_transform.xform(dest)
	
	var param = Physics2DShapeQueryParameters.new()
	param.collision_layer = $"..".collision_layer
	param.set_shape(self.shape)
	param.motion = (dest - origin)
	param.transform = self.transform
	#param.exclude = [self, $".."]
	
	var dir = space_state.cast_motion(param)
	
	var vec = dest - origin
	collision_safe = origin
	if not dir.empty():
		vec *= dir[0]
		collision_safe = origin + vec
	
	angle += 10 * delta
	
	update()
	pass
	
func _draw():
	# Draw some graphics
	draw_line(origin, dest, Color(255, 0, 0))
	draw_line(origin, collision_safe, Color(0, 255, 0), 2)
```

**Minimal reproduction project:**
CastMotionRoot.tscn inside d2 asset folder.
[godot-learn.zip](https://github.com/godotengine/godot/files/1869963/godot-learn.zip)

Best regards.

**Comment by groud** [comment]
Is this a bug or just a question ?

Here is the documentation of cast_motion:

> Checks how far the shape can travel toward a point. Note that both the shape and the motion are supplied through a Physics2DShapeQueryParameters object. The method will return an array with two floats between 0 and 1, both representing a fraction of motion. The first is how far the shape can move without triggering a collision, and the second is the point at which a collision will occur. If no collision is detected, the returned array will be 1, 1.
> 
> If the shape can not move, the array will be empty (dir.empty()==true).

It seems pretty clear to me. What's is the problem in your situation ?

**Comment by finalherizo** [comment]
Hello,

I's more a question but perhaps a bug too as I'm not able to achieve any acceptable behavior. I think I did it like what's mentioned in the documentation.
I created a Physics2DShapeQueryParameters, filled the collision layer, the shape, the motion (vector towards the destination). But it never reports a hit dir is always [1, 1].

I really don't know if I passed the right parameters. You can see what I mean inside the test project. Green is how far the shape should be able to travel. It's always green even inside the other shape.

As I said I'm perhaps using the wrong parameters so excuse me if I'm wrong.

Best regards.

**Comment by homer666** [comment]
Yeah, this isn't a bug. Where you set the transform of the query parameters (`param.transform = self.transform`), you need to supply the transform of the body itself rather than the transform of the CollisionShape2D node (so that line should be `param.transform = $"..".transform` instead). With this change, your code works correctly.

Just as a heads-up, the CollisionBody nodes are just helpers and really shouldn't be in possession of this kind of collision-checking code (I don't really know the specifics, sorry); it should ideally be in a script attached to the actual body (in this case, the `KinematicBody2D` node the `CollisionBody2D` is a child of).

**Comment by groud** [comment]
Ok, closing as it's not an engine problem then. :)

**Comment by finalherizo** [comment]
Yes, it works. Thanks for your time and sorry for that false issue.

> Just as a heads-up, the CollisionBody nodes are just helpers and really shouldn't be in possession of this kind of collision-checking code (I don't really know the specifics, sorry); it should ideally be in a script attached to the actual body (in this case, the KinematicBody2D node the CollisionBody2D is a child of).

Yes thanks. This setup is for testing purposes only. I did like what you said the first time, but I suspected that there was some problem with how I passed the transform so I put the script directly there.

I think what I didn't understand in the doc (Physics2DShapeQueryParameters) is :

> Transform2D transform - the transform matrix of the queried shape.

so I passed the CollisionShape every time (It's more clear now :))

Thanks again.

Best regards.

**Comment by homer666** [comment]
> I think what I didn't understand in the doc (Physics2DShapeQueryParameters) is :
>
> > Transform2D transform - the transform matrix of the queried shape.
>
> so I passed the CollisionShape every time (It's more clear now :))

Ah, I can see how you got confused. The docs probably need some clarification, though I'm not really sure how. Glad you fixed your problem, in any case.

**Comment by berarma** [comment]
Funny, I had to use the global_transform of the shape, not the parent, or it didn't work. There's little examples about using this calls so I'm commenting in case someone else has problems getting the collision detection to work.

My shape object had an offset and the collision detection didn't account for it unless I passed the transform from the shape. It seems the documentation was correct after all or I totally missing something and still it works.

I'm using Godot 3.4-beta4.
